### PR TITLE
Fix invokedynamic resolution for OJDK MHs

### PR DESCRIPTION
**Pull request by babsingh**
For `invokedynamic`, the first thread wins and gets to update the
`result`. The remaining threads use the `result` from the first thread.

In case of an `exception`,
- always set the the `result` to the `exception`; and
- use `CAS` to update the `callSite`.

Co-authored-by: Jack Lu <Jack.S.Lu@ibm.com>
Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
@DanHeidinga request for review.
fyi @tajila @fengxue-IS @gacholio

File: runtime/vm/resolvesupport.cpp [hunk 1]
```diff
@@ -2297,17 +2297,18 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
...
+			result = vmThread->currentException;
+		}
 
-			/* Ensure that result array elements are written before the array reference is stored */
-			VM_AtomicSupport::writeBarrier();
-			if (0 == gcFuncs->j9gc_objaccess_staticCompareAndSwapObject(vmThread, j9class, callSite, NULL, result)) {
-				/* Another thread beat this thread to updating the call site, ensure both threads return the same method handle. */
-				result = *callSite;
-			}
+		/* Only write the value in if it is not null */
```
Comment (DanHeidinga) [inline runtime/vm/resolvesupport.cpp:2303]: Delete the comment as this doesn't apply any more - we're writing something, the array or exception, provided no one else has beaten us to it.

```suggestion
```

Comment (babsingh) [reply runtime/vm/resolvesupport.cpp]: updated.

**Comment by babsingh** [comment]
Rewrote code to invoke the `writeBarrier` only if the `result` is an `array`.

File: runtime/vm/resolvesupport.cpp [hunk 2]
```diff
@@ -2297,17 +2297,21 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
 			result = vmThread->currentException;
 		} else if (result == NULL) {
```
Comment (gacholio) [inline runtime/vm/resolvesupport.cpp:2298]: The comparisons here should be inverted.

Comment (babsingh) [reply runtime/vm/resolvesupport.cpp]: inverted.

File: runtime/vm/resolvesupport.cpp [hunk 3]
```diff
@@ -2295,19 +2295,23 @@ resolveInvokeDynamic(J9VMThread *vmThread, J9ConstantPool *ramCP, UDATA callSite
...
-			J9MemoryManagerFunctions *gcFuncs = vmThread->javaVM->memoryManagerFunctions;
-			J9Class *j9class = J9_CLASS_FROM_CP(ramCP);
-
-			/* Ensure that result array elements are written before the array reference is stored */
+			result = vmThread->currentException;
+		} else {
+			/* If the result is an array, ensure that the result array elements are written
+			 * before the array reference is stored.
+			 */
 			VM_AtomicSupport::writeBarrier();
```
Comment (DanHeidinga) [inline runtime/vm/resolvesupport.cpp:2305]: I'd rather have the write barrier unconditionally as it ensures that the exception object is safely published as well.  Given that we'll rarely take the resolve path, an extra writebarrier here is fine.

Comment (babsingh) [reply runtime/vm/resolvesupport.cpp]: Updated to invoke write barrier unconditionally.

**Review by DanHeidinga — APPROVED** [review]
lgtm

**Comment by DanHeidinga** [comment]
Jenkins compile win jdk16

**Comment by DanHeidinga** [comment]
jenkins test sanity winojdk292,xlinuxojdk292 jdk16 depends ibmruntimes/openj9-openjdk-jdk16#29

**Comment by babsingh** [comment]
@DanHeidinga For the PR builds, all failures are preexisting and documented in https://github.com/eclipse/openj9/labels/project%3AMH. The Github issues for the known failures start with `OJDK MH: ...`.

**Pull request merged**